llvm/include/llvm/MC/MCRegisterInfo.h

```diff
@@ -187,6 +187,9 @@ class MCRegisterInfo {
   DenseMap<MCRegister, int> L2SEHRegs;        // LLVM to SEH regs mapping
   DenseMap<MCRegister, int> L2CVRegs;         // LLVM to CV regs mapping
 
+  mutable std::vector<std::vector<MCPhysReg>> RegAliasesCache;
```

Comment:
note: I wanted to use a vector of `unique_ptr` but it makes it too awkward to use. I can't even use `reserve` in that case, so I didn't do that indirection in the end.

The vector is resized exactly once upon init, and each vector is 3 pointers, so for AMDGPU we have 9000*24 bytes so about 216KB for this map, which is worth it IMO.

Reply:
You could use std::optional in place of unique_ptr.
Sentinel looks fine, too.

Reply:
@nikic SmallVector is 32 bytes, it has one more pointer for some reason.
Though it means I can use `SmallVector<MCPhysReg, 4>` and store small alias sets inline, so I think it's worth the size increase.

Reply:
Ah, but SmallVector doesn't have a shrink_to_fit, so it may end up using quite a bit more memory. I don't think it's worth it then.

Reply:
> @nikic SmallVector is 32 bytes, it has one more pointer for some reason.

That doesn't sound right. SmallVector should be 16 bytes large if the element type is 4 bytes... uh, I just saw that MCPhyReg is only 2 bytes, which increases SmallVector size to 24 bytes. And of course there's no way to disable that heuristic. So yeah, SmallVector won't make things smaller here (but it also shouldn't be 32 bytes either...)

> Though it means I can use `SmallVector<MCPhysReg, 4>` and store small alias sets inline, so I think it's worth the size increase.

But this also means that you allocate those 4 inline elements for all of the registers that are not actually queried. It's probably not worthwhile in this context.

Reply:
> SmallVector should be 16 bytes large if the element type is 4 bytes... uh, I just saw that MCPhyReg is only 2 bytes, which increases SmallVector size to 24 bytes.

Why does it get bigger when the element type gets smaller? Is it worth fixing that?

Reply:
This is because `SmallVector<char>` sometimes needs to hold more than 4GB of data. Use of 64-bit sizes should probably be limited to just that case and not also extend to `SmallVector<uint16_t>`...

Reply:
#95536